src/util/numerics.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+
+import * as C from "./combo";
+
+/**
+ * A library for making type-level assertions about numeric properties;
+ * e.g. that a number is finite, or finite and non-negative, or an integer.
+ *
+ * We should avoid having these methods in numeric hot loops by doing "unsafe"
+ * operations in the hot loop, and then validating on exiting the loop.
+ *
+ * Nonetheless, we avoid unnecessary function calls for performance reasons (i.e.
+ * the validator functions avoid calling each other).
+ */
+
+// A number guaranteed to not be NaN, Infinity, or -Infinity
+export opaque type Finite: number = number;
+export function finite(n: number): Finite {
+  if (!isFinite(n)) {
+    throw new Error(`expected finite, got: ${n}`);
+  }
+  return n;
+}
+export const finiteParser: C.Parser<Finite> = C.fmap(C.number, finite);
+
+export opaque type FiniteNonnegative: number = number;
+export function finiteNonnegative(n: number): FiniteNonnegative {
+  if (n < 0 || !isFinite(n)) {
+    throw new Error(`expected finite nonnegative, got: ${n}`);
+  }
+  return n;
+}
+export const finiteNonnegativeParser: C.Parser<FiniteNonnegative> = C.fmap(
+  C.number,
+  finiteNonnegative
+);
+
+export opaque type Integer: number = number;
+export function integer(n: number): Integer {
+  if (!isFinite(n) || n !== Math.floor(n)) {
```

Comment:
Can we use `Number.isSafeInteger` here? It’s hard to think of cases
where you want an argument to be an integer but don’t care whether it’s
a safe integer. For instance, naïve code might try to represent a 64-bit
uint as a JavaScript number, which would lose precision for 99.95% of
values; we should catch that. Or, from another perspective: any type
that does not have an injective “successor” operation consistent with
the “add 1” operation should not be called an integer type.

If you actually need such large integers, you should use bigints or
similar.

(also, fyi: `Number.isInteger`.)